File: app/src/main/java/com/example/proyeksp/ui/MainActivity.java
package com.example.proyeksp.ui;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.splashscreen.SplashScreen;
import androidx.lifecycle.ViewModelProvider;

import com.example.proyeksp.R;
import com.example.proyeksp.helper.CurrencyHelper;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    Button btImport, btScan, btViewData, btExport;
    TextView tvScanCount, tvTotalSetoran;
    RekeningViewModel rekViewModel;
    private static final int PERMISSION_CODE = 1001;
    ActivityResultLauncher<Intent> exportCSVLauncher, importCSVLauncher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        SplashScreen splashScreen = SplashScreen.installSplashScreen(this);
        setContentView(R.layout.activity_main);

        rekViewModel = new ViewModelProvider(this).get(RekeningViewModel.class);

        btImport = findViewById(R.id.bt_import);
        btScan = findViewById(R.id.bt_scan);
        btViewData = findViewById(R.id.bt_view_data);
        btExport = findViewById(R.id.bt_export);

        tvScanCount = findViewById(R.id.tv_scan_count);
        tvTotalSetoran = findViewById(R.id.tv_total_setoran);

        btImport.setOnClickListener(this);
        btScan.setOnClickListener(this);
        btViewData.setOnClickListener(this);
        btExport.setOnClickListener(this);

        rekViewModel.getScanData().observe(this, scanCount -> tvScanCount.setText(scanCount.toString()));
        rekViewModel.getTotalSetoran().observe(this, total -> {
                try {
                    tvTotalSetoran.setText(CurrencyHelper.format(total));
                } catch (RuntimeException e) {
                    if (total == null) {
                        tvTotalSetoran.setText("Rp0");
                    } else {
                        Toast.makeText(this, "Error: "+e.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
                        tvTotalSetoran.setText("-");
                    }
                }
            }
        );

        exportCSVLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                res -> {
                    if (res.getResultCode() == Activity.RESULT_OK) {
                        Intent data = res.getData();
                        if (data != null) {
                            Uri uri = data.getData();
                            rekViewModel.exportToXls(uri);
                            rekViewModel.getSuccess().observe(this, success -> {
                                if (success)
                                    Toast.makeText(this, "Berhasil export data", Toast.LENGTH_SHORT).show();
                                else
                                    Toast.makeText(this, "Gagal export data", Toast.LENGTH_SHORT).show();
                            });
                        }
                    }
                }
        );

        importCSVLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                res -> {
                    if (res.getResultCode() == Activity.RESULT_OK) {
                        Intent data = res.getData();
                        if (data != null) {
                            Uri uri = data.getData();
                            rekViewModel.importFromXlsx(uri);
                            rekViewModel.getSuccess().observe(this, success -> {
                                if (success)
                                    Toast.makeText(this, "Berhasil import data", Toast.LENGTH_SHORT).show();
                                else
                                    Toast.makeText(this, "Gagal import data", Toast.LENGTH_SHORT).show();
                            });
                        }
                    }
                }
        );
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == btImport.getId()) {
            Intent i = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            i.addCategory(Intent.CATEGORY_OPENABLE);
            String[] mimeTypes = {"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "application/vnd.ms-excel"};
            i.setType("*/*");
            i.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
            importCSVLauncher.launch(i);
        } else if (v.getId() == btScan.getId()) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[] {Manifest.permission.CAMERA}, PERMISSION_CODE);
            } else {
                startActivity(new Intent(this, ScanActivity.class));
            }
        } else if (v.getId() == btViewData.getId()) {
            startActivity(new Intent(this, DaftarSetoranActivity.class));
        } else if (v.getId() == btExport.getId()) {
            exportCSVLauncher.launch(new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE));
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, int[] grantResults) {
        for (int r : grantResults) {
            if (r == PackageManager.PERMISSION_DENIED) {
                Toast.makeText(this, "Harap izinkan aplikasi untuk menggunakan kamera", Toast.LENGTH_SHORT).show();
                return;
            }
        }

        if (requestCode == PERMISSION_CODE) {
            startActivity(new Intent(this, ScanActivity.class));
        }

        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }
}

File: app/src/main/java/com/example/proyeksp/ui/MainActivity.kt
package com.example.proyeksp.ui

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.compose.setContent
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ManageSearch
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.paint
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.splashscreen.SplashScreen
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.proyeksp.BuildConfig
import com.example.proyeksp.R
import com.example.proyeksp.helper.CurrencyHelper
import com.example.proyeksp.ui.components.MainButton
import com.example.proyeksp.ui.theme.AppColors
import kotlinx.coroutines.launch

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val onNavigateToScan: () -> Unit = {
            val intent = Intent(this, ScanActivity::class.java)
            startActivity(intent)
        }

        setContent {
            MainScreen(onNavigateToScan)
        }
    }
}

@Composable
fun MainScreen(onNavigateToScan: () -> Unit = {}) {
    val context = LocalContext.current

    val cameraPermissionLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.RequestPermission()
    ) { isGranted: Boolean ->
        if (isGranted) {
            // Permission granted by user -> Navigate to ScanActivity
            onNavigateToScan()
        } else {
            Toast.makeText(
                context,
                "Harap izinkan aplikasi untuk menggunakan kamera",
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    fun onScanQrClicked() {
        val isCameraPermissionGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.CAMERA
        ) == PackageManager.PERMISSION_GRANTED

        if (isCameraPermissionGranted) {
            onNavigateToScan()
        } else {
            cameraPermissionLauncher.launch(Manifest.permission.CAMERA)
        }
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .paint(
                painter = painterResource(id = R.drawable.background),
                contentScale = ContentScale.FillBounds
            ),
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
        ) {
            AndroidView(
                modifier = Modifier.fillMaxWidth(),
                factory = { context ->
                    LayoutInflater.from(context).inflate(R.layout.header, null)
                },
                update = { view ->
                    // Gambar tidak muncul kalau tidak diset manual
                    val imageView = view.findViewById<ImageView>(R.id.imageView)
                    imageView.setImageResource(R.drawable.logo_bumdes)
                }
            )
            Spacer(modifier = Modifier.weight(1f))
            Column(
                modifier = Modifier.fillMaxWidth(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                MainButton(
                    text = stringResource(id = R.string.scan_qr),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = AppColors.Pink,
                    ),
                    icon = {
                        Icon(
                            imageVector = Icons.Default.QrCodeScanner,
                            contentDescription = "Scan QR",
                            tint = Color.White,
                        )
                    },
                    onClick = {
                        onScanQrClicked()
                    }
                )
                MainButton(
                    text = stringResource(id = R.string.lihat_data),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = AppColors.Green,
                    ),
                    icon = {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ManageSearch,
                            contentDescription = "Lihat Data",
                            tint = Color.White,
                        )
                    },
                    onClick = {
                        context.startActivity(Intent(context, DaftarSetoranActivity::class.java))
                    }
                )
            }
            Spacer(modifier = Modifier.weight(1f))
            AndroidView(
                modifier = Modifier.fillMaxWidth(),
                factory = { context ->
                    LayoutInflater.from(context).inflate(R.layout.footer, null)
                }
            )
        }
    }
}

@Preview(showBackground = true)
@Composable
fun MainScreenPreview() {
    MaterialTheme {
        MainScreen()
    }
}